### ISSUE TITLE
numeric instability with kl_estimator_k3

### ISSUE THREAD

**Issue by wenxueru** [OpenRLHF/OpenRLHF#525]
<img width="283" alt="Screenshot 2024-11-19 at 18 12 48" src="https://github.com/user-attachments/assets/ad3e5656-8ac4-45d5-8a7f-2e085c41d671">
<img width="277" alt="image" src="https://github.com/user-attachments/assets/009b8ce5-347c-4f23-ab05-63f1f80c5cab">

When I use `kl_estimator_k3`, sometimes abnormally large KL values appear, which can occasionally cause the training to crash.

**Comment by hijkzzz** [comment]
@zhuzilin

**Comment by tsaoyu** [comment]
insight from ChatLearn: using logits from vLLM can sometimes cause this issue, it will be mitigate a lot by let the token produced by vLLM pass through deepspeed/torch/megatron again.

**Comment by Rosenberg37** [comment]
How about
```python
def compute_approx_kl(
    log_probs: torch.Tensor,
    log_probs_base: torch.Tensor,
    action_mask: Optional[torch.Tensor] = None,
    use_kl_estimator_k3: bool = False,
) -> torch.Tensor:
    """
    Compute the approximate KL divergence between two distributions.
    Schulman blog: http://joschu.net/blog/kl-approx.html

    Args:
        log_probs: Log probabilities of the new distribution.
        log_probs_base: Log probabilities of the base distribution.
        action_mask: Mask for actions.
    """

    log_ratio = log_probs - log_probs_base
    if action_mask is not None:
        log_ratio = log_ratio * action_mask

    # The k3 estimator is the non negative kl approximation in
    # http://joschu.net/blog/kl-approx.html
    # Besides non negative, it is also unbiased and have lower variance.
    if use_kl_estimator_k3:
        log_ratio = -log_ratio
        k3_log_ratio = log_ratio.exp() - 1 - log_ratio
        log_ratio = torch.where(log_ratio < 0, k3_log_ratio, torch.min(log_ratio, k3_log_ratio))

    return log_ratio
    ```

**Comment by wenxueru** [comment]
> How about
> 
> ```python
> def compute_approx_kl(
>     log_probs: torch.Tensor,
>     log_probs_base: torch.Tensor,
>     action_mask: Optional[torch.Tensor] = None,
>     use_kl_estimator_k3: bool = False,
> ) -> torch.Tensor:
>     """
>     Compute the approximate KL divergence between two distributions.
>     Schulman blog: http://joschu.net/blog/kl-approx.html
> 
>     Args:
>         log_probs: Log probabilities of the new distribution.
>         log_probs_base: Log probabilities of the base distribution.
>         action_mask: Mask for actions.
>     """
> 
>     log_ratio = log_probs - log_probs_base
>     if action_mask is not None:
>         log_ratio = log_ratio * action_mask
> 
>     # The k3 estimator is the non negative kl approximation in
>     # http://joschu.net/blog/kl-approx.html
>     # Besides non negative, it is also unbiased and have lower variance.
>     if use_kl_estimator_k3:
>         log_ratio = -log_ratio
>         k3_log_ratio = log_ratio.exp() - 1 - log_ratio
>         log_ratio = torch.where(log_ratio < 0, k3_log_ratio, torch.min(log_ratio, k3_log_ratio))
> 
>     return log_ratio
>     ```
> ```

Seems work to me for both non-negativity and numerical stability.

**Comment by drxmy** [comment]
> > How about
> > def compute_approx_kl(
> >     log_probs: torch.Tensor,
> >     log_probs_base: torch.Tensor,
> >     action_mask: Optional[torch.Tensor] = None,
> >     use_kl_estimator_k3: bool = False,
> > ) -> torch.Tensor:
> >     """
> >     Compute the approximate KL divergence between two distributions.
> >     Schulman blog: http://joschu.net/blog/kl-approx.html
> > 
> >     Args:
> >         log_probs: Log probabilities of the new distribution.
> >         log_probs_base: Log probabilities of the base distribution.
> >         action_mask: Mask for actions.
> >     """
> > 
> >     log_ratio = log_probs - log_probs_base
> >     if action_mask is not None:
> >         log_ratio = log_ratio * action_mask
> > 
> >     # The k3 estimator is the non negative kl approximation in
> >     # http://joschu.net/blog/kl-approx.html
> >     # Besides non negative, it is also unbiased and have lower variance.
> >     if use_kl_estimator_k3:
> >         log_ratio = -log_ratio
> >         k3_log_ratio = log_ratio.exp() - 1 - log_ratio
> >         log_ratio = torch.where(log_ratio < 0, k3_log_ratio, torch.min(log_ratio, k3_log_ratio))
> > 
> >     return log_ratio
> >     ```
> 
> Seems work to me for both non-negativity and numerical stability.

how does this affect the final model performance?

**Comment by wenxueru** [comment]
> > > How about
> > > def compute_approx_kl(
> > > log_probs: torch.Tensor,
> > > log_probs_base: torch.Tensor,
> > > action_mask: Optional[torch.Tensor] = None,
> > > use_kl_estimator_k3: bool = False,
> > > ) -> torch.Tensor:
> > > """
> > > Compute the approximate KL divergence between two distributions.
> > > Schulman blog: http://joschu.net/blog/kl-approx.html
> > > ```
> > > Args:
> > >     log_probs: Log probabilities of the new distribution.
> > >     log_probs_base: Log probabilities of the base distribution.
> > >     action_mask: Mask for actions.
> > > """
> > > 
> > > log_ratio = log_probs - log_probs_base
> > > if action_mask is not None:
> > >     log_ratio = log_ratio * action_mask
> > > 
> > > # The k3 estimator is the non negative kl approximation in
> > > # http://joschu.net/blog/kl-approx.html
> > > # Besides non negative, it is also unbiased and have lower variance.
> > > if use_kl_estimator_k3:
> > >     log_ratio = -log_ratio
> > >     k3_log_ratio = log_ratio.exp() - 1 - log_ratio
> > >     log_ratio = torch.where(log_ratio < 0, k3_log_ratio, torch.min(log_ratio, k3_log_ratio))
> > > 
> > > return log_ratio
> > > ```
> > > ```
> > 
> > 
> > Seems work to me for both non-negativity and numerical stability.
> 
> how does this affect the final model performance?

I haven't done rigorous comparative testing, but in my experiments, I found that everything works fine.

**Comment by wenxueru** [comment]
> > > > How about
> > > > def compute_approx_kl(
> > > > log_probs: torch.Tensor,
> > > > log_probs_base: torch.Tensor,
> > > > action_mask: Optional[torch.Tensor] = None,
> > > > use_kl_estimator_k3: bool = False,
> > > > ) -> torch.Tensor:
> > > > """
> > > > Compute the approximate KL divergence between two distributions.
> > > > Schulman blog: http://joschu.net/blog/kl-approx.html
> > > > ```
> > > > Args:
> > > >     log_probs: Log probabilities of the new distribution.
> > > >     log_probs_base: Log probabilities of the base distribution.
> > > >     action_mask: Mask for actions.
> > > > """
> > > > 
> > > > log_ratio = log_probs - log_probs_base
> > > > if action_mask is not None:
> > > >     log_ratio = log_ratio * action_mask
> > > > 
> > > > # The k3 estimator is the non negative kl approximation in
> > > > # http://joschu.net/blog/kl-approx.html
> > > > # Besides non negative, it is also unbiased and have lower variance.
> > > > if use_kl_estimator_k3:
> > > >     log_ratio = -log_ratio
> > > >     k3_log_ratio = log_ratio.exp() - 1 - log_ratio
> > > >     log_ratio = torch.where(log_ratio < 0, k3_log_ratio, torch.min(log_ratio, k3_log_ratio))
> > > > 
> > > > return log_ratio
> > > > ```
> > > 
> > > 
> > > Seems work to me for both non-negativity and numerical stability.
> > 
> > 
> > how does this affect the final model performance?
> 
> I haven't done rigorous comparative testing, but in my experiments, I found that everything works fine and the numerical instability issues resolved.

should be:
```
k1_log_ratio = -log_ratio
k3_log_ratio = log_ratio.exp() - 1 - log_ratio
log_ratio = torch.where(k1_log_ratio < 0, k3_log_ratio, torch.min(k1_log_ratio, k3_log_ratio))
```
